aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs

```diff
@@ -138,98 +166,80 @@ pub(crate) fn is_valid_txn_arg<S: MoveResolverExt>(
     }
 }
 
-// Validate arguments. Walk through the arguments and according to the signature
-// validate arguments that require so.
-// TODO: This needs a more solid story and a tighter integration with the VM.
-// Validation at the moment is only for Strings and Vector of them, so we
-// manually walk the serialized args until we find a string.
-// This is obviously brittle and something to change at some point soon.
-pub(crate) fn validate_args<S: MoveResolverExt>(
-    session: &SessionExt<S>,
-    idxs: &[usize],
-    args: &[Vec<u8>],
-    func: &LoadedFunctionInstantiation,
-) -> Result<(), VMStatus> {
-    for (idx, (ty, arg)) in func.parameters.iter().zip(args.iter()).enumerate() {
-        if !idxs.contains(&idx) {
-            continue;
-        }
-        let arg_len = arg.len();
-        let mut cursor = Cursor::new(&arg[..]);
-        validate_arg(session, ty, &mut cursor, arg_len)?;
+fn validate_and_construct<S: MoveResolverExt>(
+    session: &mut SessionExt<S>,
+    expected_type: &Type,
+    constructor: &FunctionId,
+    cursor: &mut Cursor<&[u8]>,
+    gas_meter: &mut impl GasMeter,
+) -> Result<Vec<u8>, VMStatus> {
+    let (module, function, instantiation) = session.load_function_with_type_arg_inference(
+        &constructor.module_id,
+        IdentStr::new(constructor.func_name).expect(""),
+        expected_type,
+    )?;
+    let mut args = vec![];
+    for param_type in &instantiation.parameters {
+        let mut arg = vec![];
+        recursively_construct_arg(session, param_type, cursor, gas_meter, &mut arg)?;
+        args.push(arg);
     }
-    Ok(())
+    let serialized_result = session
+        .execute_instantiated_function(module, function, instantiation, args, gas_meter)
+        .map_err(|_| VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT, None))?;
+    let mut ret_vals = serialized_result.return_values;
+    // We know ret_vals.len() == 1
+    Ok(ret_vals.pop().expect("Always a result").0)
 }
 
-// Validate a single arg. A Cursor is used to walk the serialized arg manually and correctly.
-// Only Strings and nested vector of them are validated.
-fn validate_arg<S: MoveResolverExt>(
-    session: &SessionExt<S>,
+// A Cursor is used to recursively walk the serialized arg manually and correctly.
+pub(crate) fn recursively_construct_arg<S: MoveResolverExt>(
+    session: &mut SessionExt<S>,
     ty: &Type,
     cursor: &mut Cursor<&[u8]>,
-    arg_len: usize,
+    gas_meter: &mut impl GasMeter,
+    arg: &mut Vec<u8>,
 ) -> Result<(), VMStatus> {
     use move_vm_types::loaded_data::runtime_types::Type::*;
 
     match ty {
         Vector(inner) => {
             // get the vector length and iterate over each element
             let mut len = get_len(cursor)?;
+            serialize_uleb128(len, arg);
```

Comment:
What is this for?

Reply:
When you have a vector of move structs, then we need to construct each of the move structs and reserialize them, which is what this code does. Ie. it BCS encodes the length of the vector and then recursively BCS encodes each struct

Reply:
Personally I don't like this one cuz it didn't reuse bcs code.
I wonder can you call `ser::Serializer::serialize_seq(Some(len)` instead?

Reply:
getlen and the parsing of the previous validation function is already exposing the BCS format directly, so reserializing is not adding much

Reply:
Yeah in this particular case I don't actually think there's an easy way to reuse the BCS code. 

The reason is that we're effectively injecting a format change into the value representation, with struct being represented as a list of arguments to the constructor instead of its fields. Because values are recursive, the whole parsing/reserialization process need to be rewritten.